File: Autonomous_Mobile_Robots_Course/Chapter6/Lesson4/Chapter6_Lesson4.java
// Chapter6_Lesson4.java
// Independence Assumptions and Their Limits — correlated-sensor Bayesian fusion (Java)
//
// Demonstrates overconfidence when correlation is ignored.
//
// Libraries: none required (pure Java). For larger-scale robotics estimation,
// you would typically integrate with rosjava or Apache Commons Math.
// Compile:
//   javac Chapter6_Lesson4.java
// Run:
//   java Chapter6_Lesson4

public class Chapter6_Lesson4 {

    static class Posterior {
        public double mean;
        public double var;
        Posterior(double mean, double var) { this.mean = mean; this.var = var; }
    }

    // Inverse of Sigma = sigma^2 [[1,rho],[rho,1]]
    static double[][] invSigma(double sigma, double rho) {
        double s2 = sigma * sigma;
        double det = s2 * s2 * (1.0 - rho * rho);
        double[][] inv = new double[2][2];
        inv[0][0] =  s2 / det;   // 1/(sigma^2*(1-rho^2))
        inv[0][1] = -rho*s2 / det;
        inv[1][0] = -rho*s2 / det;
        inv[1][1] =  s2 / det;
        return inv;
    }

    static Posterior posteriorCorrelated(double mu0, double sigma0, double y1, double y2, double sigma, double rho) {
        double[][] Sinv = invSigma(sigma, rho);

        // H = [1,1]^T. info_meas = H^T Sinv H = sum of all entries of Sinv
        double info_meas = Sinv[0][0] + Sinv[0][1] + Sinv[1][0] + Sinv[1][1];

        double postVar = 1.0 / (1.0/(sigma0*sigma0) + info_meas);

        // info_vec = H^T Sinv y
        double info_vec = (Sinv[0][0] + Sinv[1][0]) * y1 + (Sinv[0][1] + Sinv[1][1]) * y2;
        double postMean = postVar * (mu0/(sigma0*sigma0) + info_vec);

        return new Posterior(postMean, postVar);
    }

    static Posterior posteriorIndependent(double mu0, double sigma0, double y1, double y2, double sigma) {
        double postVar = 1.0 / (1.0/(sigma0*sigma0) + 2.0/(sigma*sigma));
        double postMean = postVar * (mu0/(sigma0*sigma0) + (y1 + y2)/(sigma*sigma));
        return new Posterior(postMean, postVar);
    }

    public static void main(String[] args) {
        double mu0 = 0.0;
        double sigma0 = 2.0;
        double sigma = 1.0;
        double y1 = 1.0, y2 = 1.2;

        System.out.println(String.format("Prior: x ~ N(mu0, sigma0^2), mu0=%.6f, sigma0=%.6f", mu0, sigma0));
        System.out.println(String.format("Measurements: y1=%.6f, y2=%.6f, sigma=%.6f\n", y1, y2, sigma));

        double[] rhos = new double[] {0.0, 0.3, 0.6, 0.9};
        for (double rho : rhos) {
            Posterior pc = posteriorCorrelated(mu0, sigma0, y1, y2, sigma, rho);
            Posterior pi = posteriorIndependent(mu0, sigma0, y1, y2, sigma);
            double ratio = pi.var / pc.var;

            System.out.println(String.format("rho=%.1f:", rho));
            System.out.println(String.format("  Correct correlated posterior: mean=%.6f, var=%.6f", pc.mean, pc.var));
            System.out.println(String.format("  Indep. assumption posterior:  mean=%.6f, var=%.6f", pi.mean, pi.var));
            System.out.println(String.format("  Variance ratio (indep/correct) = %.6f ( < 1 means overconfident )\n", ratio));
        }
    }
}

File: Autonomous Mobile Robots Course/Chapter6/Lesson4/Chapter6_Lesson4.cpp
// Chapter6_Lesson4.cpp
// Independence Assumptions and Their Limits — correlated-sensor Bayesian fusion (C++)
//
// Demonstrates overconfidence when measurement correlation is ignored.
//
// Dependencies: Eigen3 (common in robotics).
// Compile (example):
//   g++ -O2 -std=c++17 Chapter6_Lesson4.cpp -I /usr/include/eigen3 -o ch6l4
// Run:
//   ./ch6l4

#include <iostream>
#include <iomanip>
#include <Eigen/Dense>

struct Posterior {
  double mean;
  double var;
};

Posterior posterior_correlated(double mu0, double sigma0, const Eigen::Vector2d& y, double sigma, double rho) {
  Eigen::Matrix2d Sigma;
  Sigma << 1.0, rho,
           rho, 1.0;
  Sigma *= (sigma * sigma);

  Eigen::Matrix2d Sinv = Sigma.inverse();
  Eigen::Vector2d H;
  H << 1.0, 1.0;

  double info_meas = H.transpose() * Sinv * H;                 // scalar
  double post_var = 1.0 / (1.0/(sigma0*sigma0) + info_meas);   // scalar

  double info_vec = H.transpose() * Sinv * y;                  // scalar
  double post_mean = post_var * (mu0/(sigma0*sigma0) + info_vec);

  return {post_mean, post_var};
}

Posterior posterior_independent(double mu0, double sigma0, const Eigen::Vector2d& y, double sigma) {
  // rho=0 -> Sigma = sigma^2 I; precision additivity
  double post_var = 1.0 / (1.0/(sigma0*sigma0) + 2.0/(sigma*sigma));
  double post_mean = post_var * (mu0/(sigma0*sigma0) + (y(0) + y(1))/(sigma*sigma));
  return {post_mean, post_var};
}

int main() {
  double mu0 = 0.0;
  double sigma0 = 2.0;
  double sigma = 1.0;
  Eigen::Vector2d y(1.0, 1.2);

  std::cout << std::fixed << std::setprecision(6);
  std::cout << "Prior: x ~ N(mu0, sigma0^2) with mu0=" << mu0 << ", sigma0=" << sigma0 << "\n";
  std::cout << "Measurements: y1=" << y(0) << ", y2=" << y(1) << ", sigma=" << sigma << "\n\n";

  for (double rho : {0.0, 0.3, 0.6, 0.9}) {
    Posterior pc = posterior_correlated(mu0, sigma0, y, sigma, rho);
    Posterior pi = posterior_independent(mu0, sigma0, y, sigma);

    double ratio = pi.var / pc.var;
    std::cout << "rho=" << rho << ":\n";
    std::cout << "  Correct correlated posterior: mean=" << pc.mean << ", var=" << pc.var << "\n";
    std::cout << "  Indep. assumption posterior:  mean=" << pi.mean << ", var=" << pi.var << "\n";
    std::cout << "  Variance ratio (indep/correct) = " << ratio << " ( < 1 means overconfident )\n\n";
  }
  return 0;
}
